src/blockMonitor/blockProcessor.ts

```diff
@@ -66,6 +82,31 @@ export class BlockProcessor extends StartStopService {
         this.provider.removeListener("block", this.processBlockNumber);
     }
 
+    // update the new headHash if needed, and emit the appropriate events
+    private processNewHead(headBlock: ethers.providers.Block, commonAncestorBlock: IBlockStub | null) {
+        const oldHeadHash = this.headHash; // we need to remember the old head for proper Reorg event handling
+        this.headHash = headBlock.hash;
+
+        if (!this.headHash || this.headHash === this.lastBlockHashReceived) {
+            // Emit the appropriate events, but only if the service is already started
+            if (this.started) {
+                if (
+                    !commonAncestorBlock || // deep reorg, no common ancestor
```

Comment:
To make this a bit clearer, you could split this check in two, something like:
```
if (!commonAncestorBlock){
             // maybe here you emit catastrophic reorg? or just REORG with different parameters
             // I suppose this depends on how we want to use catastrophic reorg? we should probably just continue
             // but also raise the alarm! (having said that, catastrophic reorg can occur quite often when the cache is just
             // beginning to be populated
             this.emit(
                        BlockProcessor.REORG_EVENT,
                        null,
                        this.headHash,
                        oldHeadHash
                    );
}
else if(oldHeadHash !== commonAncestorBlock.hash) {
                this.emit(
                        BlockProcessor.REORG_EVENT,
                        commonAncestorBlock.hash,
                        this.headHash,
                        oldHeadHash
                    );
                }

// else do nothing - this is just an extension of the head
```

Reply:
Done in a0bdd08, much cleaner indeed.

About the comment: since the listeners can easily distinguish between a catastrophic reorg and a normal reorg (by checking the commonAncestor parameter), I would prefer keep just one event in the BlockProcessor, and leave that to the business logic of the listener.
Anyway, we might decide to refactor when we have different components using these events.